minidump-common/src/format.rs

```diff
@@ -1389,33 +1389,74 @@ multi_structs! {
+    pub struct MINIDUMP_MISC_INFO5 {
+        pub xstate_data: XSTATE_CONFIG_FEATURE_MSC_INFO,
+        pub process_cookie: u32,
+    }
+    // TODO?: read xstate_data.enabled_features and process the extra XSAVE
+    // sections that should follow MISC_INFO_5 entries as documented in
+    // Volume 1, Chapter 13 of the "Intel 64 and IA-32 Architectures
+    // Software Developer’s Manual".
 }
 
-//TODO: MINIDUMP_MISC_INFO_5
-/*
-typedef struct _MINIDUMP_MISC_INFO_5 {
-    ULONG32 SizeOfInfo;
-    ULONG32 Flags1;
-    ULONG32 ProcessId;
-    ULONG32 ProcessCreateTime;
-    ULONG32 ProcessUserTime;
-    ULONG32 ProcessKernelTime;
-    ULONG32 ProcessorMaxMhz;
-    ULONG32 ProcessorCurrentMhz;
-    ULONG32 ProcessorMhzLimit;
-    ULONG32 ProcessorMaxIdleState;
-    ULONG32 ProcessorCurrentIdleState;
-    ULONG32 ProcessIntegrityLevel;
-    ULONG32 ProcessExecuteFlags;
-    ULONG32 ProtectedProcess;
-    ULONG32 TimeZoneId;
-    TIME_ZONE_INFORMATION TimeZone;
-    WCHAR   BuildString[MAX_PATH];
-    WCHAR   DbgBldStr[40];
-    XSTATE_CONFIG_FEATURE_MSC_INFO XStateData;
-    ULONG32 ProcessCookie;
-} MINIDUMP_MISC_INFO_5, *PMINIDUMP_MISC_INFO_5;
-*/
+#[derive(Clone, Pread, SizeWith)]
+pub struct XSTATE_CONFIG_FEATURE_MSC_INFO {
+    pub size_of_info: u32,
+    pub context_size: u32,
+    // features[i] is only valid if enabled_features[i] is set
+    pub enabled_features: XStateFeatures,
+    pub features: [XSTATE_FEATURE; 64],
+}
+
+#[repr(usize)]
+pub enum XStateFeatureIndex {
+    LEGACY_FLOATING_POINT = 0,
+    LEGACY_SSE = 1,
+    GSSE_AND_AVX = 2,
+    MPX_BNDREGS = 3,
+    MPX_BNDCSR = 4,
+    AVX512_KMASK = 5,
+    AVX512_ZMM_H = 6,
+    ACK512_ZMM = 7,
+    XSTATE_IPT = 8,
+    XSTATE_LWP = 62,
+}
+
+bitflags! {
+    #[derive(Pread, SizeWith)]
+    pub struct XStateFeatures: u64 {
+        const XSTATE_LEGACY_FLOATING_POINT  = 1 << (XStateFeatureIndex::LEGACY_FLOATING_POINT as usize);
+        const XSTATE_LEGACY_SSE             = 1 << (XStateFeatureIndex::LEGACY_SSE as usize);
+        const XSTATE_GSSE                   = 1 << (XStateFeatureIndex::GSSE_AND_AVX as usize);
+        const XSTATE_AVX                    = 1 << (XStateFeatureIndex::GSSE_AND_AVX as usize);
+        const XSTATE_MPX_BNDREGS            = 1 << (XStateFeatureIndex::MPX_BNDREGS as usize);
+        const XSTATE_MPX_BNDCSR             = 1 << (XStateFeatureIndex::MPX_BNDCSR as usize);
+        const XSTATE_AVX512_KMASK           = 1 << (XStateFeatureIndex::AVX512_KMASK as usize);
+        const XSTATE_AVX512_ZMM_H           = 1 << (XStateFeatureIndex::AVX512_ZMM_H as usize);
+        const XSTATE_AVX512_ZMM             = 1 << (XStateFeatureIndex::ACK512_ZMM as usize);
+        const XSTATE_IPT                    = 1 << (XStateFeatureIndex::XSTATE_IPT as usize);
+        const XSTATE_LWP                    = 1 << (XStateFeatureIndex::XSTATE_LWP as usize);
+    }
+}
+
+#[derive(Clone, Copy, Pread, SizeWith)]
+pub struct XSTATE_FEATURE {
+    pub offset: u32,
```

Comment:
It would be great to assume some of Crashpad's doc comments:

- https://crashpad.chromium.org/doxygen/structXSTATE__FEATURE.html
- https://crashpad.chromium.org/doxygen/structXSTATE__CONFIG__FEATURE__MSC__INFO.html

Reply:
I made my own versions since these refer a lot to their own APIs